tests/e2e/support/api/keycloak/ocisUserToken.ts

```diff
@@ -0,0 +1,141 @@
+}
+
+const authorizationEndpoint = config.keycloakUrl + '/realms/oCIS/protocol/openid-connect/auth'
+const tokenEndpoint = config.keycloakUrl + '/realms/oCIS/protocol/openid-connect/token'
+const redirectUrl = config.backendUrl + '/oidc-callback.html'
+
+async function getAuthorizationEndPoint() {
+  const loginParams = {
+    client_id: 'web',
+    redirect_uri: redirectUrl,
+    response_mode: 'query',
+    response_type: 'code',
+    scope: 'openid profile email'
+  }
+  const queryString = new URLSearchParams(loginParams).toString()
+  const authorizationUrl = `${authorizationEndpoint}?${queryString}`
+
+  const authorizationResponse = await fetch(authorizationUrl, {
+    method: 'GET',
+    redirect: 'manual'
+  })
+
+  if (authorizationResponse.status === 302) {
+    const locationHeader = authorizationResponse.headers.get('location')
+    const urlParams = new URLSearchParams(new URL(locationHeader).search)
+    const errorDescription = urlParams.get('error_description')
+    throw new Error(`Unexpected redirection. ${errorDescription}`)
+  } else if (authorizationResponse.status !== 200) {
+    throw new Error(
+      `Authorization failed: Expected status code to be 200 but received ${authorizationResponse.status}. \nMessage: ${authorizationResponse.statusText}`
+    )
+  }
+
+  const cookies = authorizationResponse.headers.raw()['set-cookie']?.[0]
+  const htmlData = await authorizationResponse.text()
+
+  // authorization url for login is send back from server in the HTML body.
+  const match = htmlData.match(/action="([^"]+)"/)
+  if (!match) {
+    throw new Error('No authorization url found in the HTML response body.')
+  }
+  const auhorizationUrl = match[1]
+  return [auhorizationUrl, cookies]
+}
+
+const getCode = async (user, auhorizationUrl: string, cookies: string) => {
+  const authCodeResponse = await fetch(auhorizationUrl, {
+    method: 'POST',
+    body: new URLSearchParams({
+      username: user.id,
+      password: user.password
+    }),
+    redirect: 'manual',
+    headers: {
+      Cookie: cookies
+    }
+  })
+
+  if (authCodeResponse.status !== 302) {
+    throw new Error(
+      `Login failed: Expected status code to be 302 but received ${authCodeResponse.status}. \nMessage: ${authCodeResponse.statusText}`
+    )
+  }
+
+  const locationHeader = authCodeResponse.headers.get('location')
+  const urlParams = new URLSearchParams(new URL(locationHeader).search)
+  return urlParams.get('code')
+}
+
+const getToken = async (authorizationCode: string) => {
+  const tokenResponse = await fetch(tokenEndpoint, {
+    method: 'POST',
+    body: new URLSearchParams({
+      client_id: 'web',
+      code: authorizationCode,
+      redirect_uri: redirectUrl,
+      grant_type: 'authorization_code'
+    })
+  })
+
+  if (tokenResponse.status !== 200) {
+    throw new Error(
+      `Failed to retrieve token: Expected status code to be 200 but received ${tokenResponse.status}. \nMessage: ${tokenResponse.statusText}`
+    )
+  }
```

Comment:
This condition with error throw is repeated
https://github.com/owncloud/web/blob/d57b7ae2ff001cfdc863fbffad2639dd1cfbda88/tests/e2e/support/api/keycloak/ocisUserToken.ts#L126-L129

Reply:
one function is to get access token while other is to get accesstoken using `refreshToken` . I don't understand why this repetition is an issue. I don't see any link